### PR TITLE
Support Ubuntu 20.04

### PR DESCRIPTION
**Pull request by treydock**
#### Pull Request (PR) description

Add support for Ubuntu 20.04.  Had to force virtualenv to not use `distribute` otherwise produces errors.  Also updated virtualenv tests to consistently use Python 3 since first set of tests did this but subsequent tests did not and that broke some OS acceptance tests.

#### This Pull Request (PR) fixes the following issues

Replaces #576

### CODE REVIEW

**Review by bastelfreak — APPROVED** [review]

**Comment by treydock** [comment]
Looks like the Ubuntu 20.02 & Puppet 5 acceptance failure was networking issues boostrapping the beaker nodeset.  Will need to be rerun.

**Comment by bastelfreak** [comment]
Ubuntu 29 + puppet 5 is expected to fail because puppet does not provide packages fro them :(

**Comment by treydock** [comment]
Looks like majority of tests are failing virtualenv with this error:

```
       	Notice: /Stage[main]/Main/Python::Virtualenv[venv]/Exec[python_virtualenv_/opt/venv6]/returns:     AttributeError: module 'importlib._bootstrap' has no attribute 'SourceFileLoader'
```

After much googling at least with Ubuntu 20.04 the online solutions did not work.  The only way around it was not trying to install `distribute` pip package so forcing `distribute => false` for `python::virtualenv`.  I'm wondering if that should become the new default because majority of OSes with Python 3 seem to not work with `distribute => true`.

File: manifests/pyvenv.pp [hunk 1]
```diff
@@ -48,7 +48,7 @@
       $python3_venv_package = "python${normalized_python_version}-venv"
 
       case $facts['os']['distro']['codename'] {
-        'xenial', 'bionic', 'cosmic', 'disco', 'stretch', 'buster': {
+        'xenial', 'bionic', 'cosmic', 'disco', 'stretch', 'buster', 'focal': {
```
Comment (bastelfreak) [inline manifests/pyvenv.pp:51]: that's all the ubuntu/debian versions we support. I think we can just drop this case statement?

**Comment by treydock** [comment]
@bastelfreak I am not real familiar with how voxpupuli modulesync sets up `setfiles` for acceptance, my exclude didn't seem to take.   I think the hash used to define matrix causes issues with being able to exclude elements

File: manifests/pyvenv.pp (hunk 1 above)
Comment (treydock) [reply manifests/pyvenv.pp]: Ya, I dropped it.

**Comment by bastelfreak** [comment]
we havn't figured out yet how to exclude sets :(

**Review by bastelfreak — APPROVED** [review]
looks good to me, but will keep it open for a day or two so others can review it as well

**Comment by bastelfreak** [comment]
@crazymind1337 can you take a look as well?

**Comment by bastelfreak** [comment]
@treydock can you rebase against master to get rid of the merge commit?

**Comment by treydock** [comment]
@bastelfreak I had done a merge to get in the virtualenv unit tests which needed to be adjusted for this pull request. Looks to have resolved the conflicts.

File: manifests/install.pp [hunk 2]
```diff
@@ -284,8 +284,11 @@
         $pip_package  = 'pip'
         $pip_provider = 'pip'
       } else {
+        $pip_package = fact('os.release.major') ? {
```
Comment (crazymind1337) [inline manifests/install.pp:287]: Can we use here the $facts-Hash? I know everybody has his favor way to access the facts. But just to be consistent within the module I would like to see the $facts-Hash.

File: manifests/params.pp [hunk 3]
```diff
@@ -46,4 +45,9 @@
     'RedHat' => 'python-gunicorn',
     default  => 'gunicorn',
   }
+
+  $version = fact('os.release.major') ? {
```
Comment (crazymind1337) [inline manifests/params.pp:49]: Same here.

File: manifests/pyvenv.pp [hunk 4]
```diff
@@ -46,15 +46,8 @@
...
-      case $facts['os']['distro']['codename'] {
-        'xenial', 'bionic', 'cosmic', 'disco', 'stretch', 'buster': {
-          ensure_packages($python3_venv_package)
-
-          Package[$python3_venv_package] -> File[$venv_dir]
-        }
-        default: {}
-      }
+      ensure_packages($python3_venv_package)
+      Package[$python3_venv_package] -> File[$venv_dir]
```
Comment (crazymind1337) [inline manifests/pyvenv.pp:50]: Could you readd the empty line between `ensure_packages` and the sorting?

File: manifests/install.pp (hunk 2 above)
Comment (treydock) [reply manifests/install.pp]: Used fact hash, `$fact['os']['release']['major']`.

File: manifests/params.pp (hunk 3 above)
Comment (treydock) [reply manifests/params.pp]: Used fact hash, `$fact['os']['release']['major']`.

File: manifests/pyvenv.pp (hunk 4 above)
Comment (treydock) [reply manifests/pyvenv.pp:51]: Added empty line

**Review by bastelfreak — APPROVED** [review]

**Comment by crazymind1337** [comment]
While testing on Ubuntu 20.04 I get this result: ```Notice: /Stage[main]/Python::Install/Package[python]/ensure: created
Error: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install python-pip' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package python-pip
Error: /Stage[main]/Python::Install/Package[pip]/ensure: change from 'purged' to 'present' failed: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install python-pip' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package python-pip```

**Comment by crazymind1337** [comment]
Tested on latest ubuntu:20.04 docker container. Works for me!

**Review by crazymind1337 — APPROVED** [review]

**Comment by crazymind1337** [comment]
@treydock You still have to fix the Commit messages. It looks like there is something wrong.

**Comment by crazymind1337** [comment]
> @treydock You still have to fix the Commit messages. It looks like there is something wrong.

Nevermind. I will squash it.

**Pull request merged**